File: tobas_setup_assistant/src/tobas_setup_assistant/setting_widgets/ros_package.py
from __future__ import annotations
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ..setup_assistant import SetupAssistant
    from ..parameter_getters import ParamGetterWidget

import os
import os.path as osp
import re
from glob import glob
from overrides import override
from PyQt5.QtCore import Qt, pyqtSlot
from PyQt5.QtWidgets import QLabel, QPushButton, QVBoxLayout
from PyQt5.QtGui import QFont

from tobas_rqt_tools.path import get_catkin_ws_path, get_catkin_ws_paths, is_in_catkin_src
from tobas_rqt_tools.utils import place_center
from tobas_rqt_tools.messages import q_error, yes_or_no, QMessageLevel
from tobas_tools_py.constants import PKG_EXTENSION

from ..common import BODY_PSIZE
from ..parameter_getters import ParamGetterWidget_DirDialog, ParamGetterWidget_LineEdit
from ..utils import get_drone_name
from .base_setting import BaseSettingWidget


class RosPackageWidget(BaseSettingWidget):
    NAME = "ROS Package"
    TITLE_TEXT = "Generate ROS Package"
    ABST_TEXT = (
        "Based on the previous settings, we will generate the necessary ROS packages for using Tobas. "
        'Please specify the path for the package and click the "Generate" button.'
    )

    TEXT_HEIGHT = 50
    BUTTON_HEIGHT = 40
    BUTTON_WIDTH = 100

    def __init__(self, main: SetupAssistant) -> None:
        super().__init__(main)

        self._param_rows = QVBoxLayout()
        self._rows.addLayout(self._param_rows)

        pardir_description = ""
        self._pardir = ParamGetterWidget_DirDialog("Parent Directory", pardir_description)
        self._pardir.path_changed.connect(self._on_path_changed)
        self._param_rows.addWidget(self._pardir)

        pkg_name_description = ""
        self._tbs_name = ParamGetterWidget_LineEdit("Package Name", pkg_name_description)
        self._tbs_name.text_changed.connect(self._on_path_changed)
        self._param_rows.addWidget(self._tbs_name)

        text = QLabel("The package will be generated as")
        text.setFont(QFont("Default", pointSize=BODY_PSIZE))
        text.setFixedHeight(self.TEXT_HEIGHT)
        self.setAlignment(Qt.AlignTop)
        self._rows.addWidget(text)

        self._tbs_path = QLabel(main)
        self._tbs_path.setFont(QFont("Default", pointSize=BODY_PSIZE, weight=QFont.Bold))
        self._tbs_path.setFixedHeight(self.TEXT_HEIGHT)
        self._tbs_path.setAlignment(Qt.AlignTop)
        self._rows.addWidget(self._tbs_path)

        # ボタンを中央に配置するためにLayoutとWidgetを噛ませる必要がある
        self._generate_button = QPushButton("Generate")
        self._generate_button.setFixedSize(self.BUTTON_WIDTH, self.BUTTON_HEIGHT)
        self._generate_button.setEnabled(False)
        self._generate_button.clicked.connect(self._on_generate_button_clicked)
        place_center(self._generate_button, self._rows)

        self._rows.addStretch()

    @override
    def update_internal_data_structures(self) -> None:
        # デフォルトのsrcディレクトリを設定
        ws_path = self._last_accessed_ws_path()
        src_path = osp.join(ws_path, "src")
        self._pardir.set(src_path)

        # デフォルトのパッケージ名を設定
        tbs_name = f"tobas_{get_drone_name()}"
        self._tbs_name.set(tbs_name)

    @override
    def is_valid(self) -> bool:
        pardir = self._pardir.get()
        tbs_name = self._tbs_name.get()
        tbs_path = self._tbs_path.text()

        # 親ディレクトリが存在することを確認
        if not osp.isdir(pardir):
            q_error(self._main, f'"{pardir}" does not exist.')
            return False
        if not is_in_catkin_src(pardir):
            q_error(self._main, f'"{pardir}" is not in the src directory of a catkin workspace.')
            return False

        # パッケージ名が無効な文字を含んでいないことを確認
        if tbs_name.count("/") > 0 or tbs_name.count(".") > 0 or tbs_name.count(" ") > 0:
            q_error(self._main, f"Invalid package name: {tbs_name}")
            return False

        # 同じcatkinワークスペースのソースディレクトリ内に同じ名前でパスが異なるTobasパッケージが存在しないことを確認
        for same_name_pkg in glob(osp.join(get_catkin_ws_path(pardir), "src", "*", self._tbs_name_with_ext())):
            if same_name_pkg != tbs_path:
                q_error(self._main, f'"{self._tbs_name_with_ext()}" already exists.: {same_name_pkg}')
                return False

        # パッケージパスが既に存在する場合は置換するかどうかをユーザに確認
        if osp.exists(tbs_path):
            if not yes_or_no(self._main, f"{tbs_path} already exists. Do you want to replace it?", QMessageLevel.WARN):
                return False

        return True

    @override
    def dump_settings(self) -> dict:
        res = dict()
        for i in range(self._param_rows.count()):
            param: ParamGetterWidget = self._param_rows.itemAt(i).widget()
            res[param.name()] = param.get()
        return res

    @override
    def load_settings(self, data: dict) -> None:
        for i in range(self._param_rows.count()):
            param: ParamGetterWidget = self._param_rows.itemAt(i).widget()
            param.set(data[param.name()])

    def tbs_name(self) -> str:
        return self._tbs_name.get()

    def tbs_path(self) -> str:
        return self._tbs_path.text()

    @pyqtSlot()
    def _on_generate_button_clicked(self) -> None:
        self._main.pkg_generator.generate_package()

    @pyqtSlot()
    def _on_path_changed(self) -> None:
        pardir = self._pardir.get()
        tbs_name = self._tbs_name.get()

        path = pardir + "/" + self._tbs_name_with_ext()
        path = re.sub("/*/", "/", path)  # スラッシュの重複を削除
        self._tbs_path.setText(path)

        self._generate_button.setEnabled(pardir != "" and tbs_name != "")

    def _last_accessed_ws_path(self) -> str:
        catkin_ws_paths = get_catkin_ws_paths()

        # もしcatkin_wsが存在しなければ作る
        if len(catkin_ws_paths) == 0:
            ws_path = osp.expanduser("~/catkin_ws/")
            src_path = osp.join(ws_path, "src")
            os.makedirs(src_path, exist_ok=True)
            os.chdir(ws_path)
            if os.system("catkin init") != 0:
                raise RuntimeError("Failed to create catkin workspace.")
            return ws_path

        cnd_ws = None
        cnd_time = 0
        for ws in catkin_ws_paths:
            last_accessed_time = osp.getatime(ws)
            if last_accessed_time > cnd_time:
                cnd_ws = ws
                cnd_time = last_accessed_time

        return cnd_ws

    def _tbs_name_with_ext(self) -> str:
        return self.tbs_name() + PKG_EXTENSION

File: tobas_gui/tobas_setup_assistant/src/setting_widgets/ros_package.cpp
#include <filesystem>

#include <tobas_linux/core.hpp>
#include <tobas_yaml_tools/convert/qstring.hpp>
#include <tobas_constants/constants.hpp>
#include <tobas_qt_tools/font.hpp>
#include <tobas_qt_tools/message.hpp>

#include "tobas_setup_assistant/setting_tabs/ros_package.hpp"

namespace fs = std::filesystem;

namespace gui
{
namespace setup_assistant
{
ROSPackageWidget::ROSPackageWidget(rclcpp::Node::SharedPtr node, const RobotInfo& robot) : node_(node), robot_(robot)
{
}

const char* ROSPackageWidget::name() const
{
  return "ROS Package";
}

const char* ROSPackageWidget::title() const
{
  return "Generate ROS Package";
}

const char* ROSPackageWidget::description() const
{
  return "Based on the previous settings, we will generate the necessary ROS packages for using Tobas. "
         "Please specify the path for the package and click the \"Generate\" button.";
}

void ROSPackageWidget::onInit()
{
  pardir_ = new ParamGetterWidget_DirDialog(node_, "Parent Directory", "");
  connect(pardir_, &ParamGetterWidget_DirDialog::pathChanged, this, &self::onPathChanged);
  addWidget(pardir_);

  tbs_name_ = new ParamGetterWidget_LineEdit("Package Name", "");
  connect(tbs_name_, &ParamGetterWidget_LineEdit::textChanged, this, &self::onPathChanged);
  addWidget(tbs_name_);

  const auto text = new QLabel("The package will be generated as");
  text->setFont(qt::DefaultFont(kBodyPSize));
  text->setFixedHeight(kTextHeight);
  setAlignment(Qt::AlignTop);
  addWidget(text);

  tbs_path_ = new QLabel();
  tbs_path_->setFont(qt::DefaultFont(kBodyPSize, QFont::Bold));
  tbs_path_->setFixedHeight(kTextHeight);
  tbs_path_->setAlignment(Qt::AlignTop);
  addWidget(tbs_path_);

  generate_button_ = new QPushButton("Generate");
  generate_button_->setFixedSize(kButtonWidth, kButtonHeight);
  generate_button_->setEnabled(false);
  connect(generate_button_, &QPushButton::clicked, this, &self::onGenerateButtonClicked);
  addWidgetCenter(generate_button_);
}

void ROSPackageWidget::onOpened()
{
  return;
}

void ROSPackageWidget::updateInternalDataStructures()
{
  // デフォルトの親ディレクトリを設定
  const auto default_pardir = path::join(linux::expandUser(tobas::kColconWSPath), "src");
  pardir_->setValue(QString::fromStdString(default_pardir));

  // デフォルトの親ディレクトリが存在しなければ作成
  if (!fs::exists(default_pardir))
    if (!fs::create_directories(default_pardir))
      qt::qErrorBox(this, QString::fromStdString("Failed to create \"" + default_pardir + "\"."));

  // デフォルトのTBSパッケージ名を設定
  const auto tbs_name = "tobas_" + robot_.robotName();
  tbs_name_->setValue(QString::fromStdString(tbs_name));
}

bool ROSPackageWidget::isValid()
{
  const auto pardir = pardir_->getValue();
  const auto tbs_name = tbs_name_->getValue();
  const auto tbs_path = tbs_path_->text();

  // 親ディレクトリが存在することを確認
  if (!fs::is_directory(pardir.toStdString()))
  {
    qt::qErrorBox(this, pardir + " does not exist.");
    return false;
  }

  // パッケージ名が無効な文字を含んでいないことを確認
  if (tbs_name.contains('/') || tbs_name.contains('.') || tbs_name.contains(' '))
  {
    qt::qErrorBox(this, "Invalid package name: " + tbs_name);
    return false;
  }

  // パッケージパスが既に存在する場合は置換するかどうかをユーザに確認
  if (fs::exists(tbs_path.toStdString()))
    if (!qt::yesOrNo(this, tbs_path + " already exists. Do you want to replace it?", qt::QMessageLevel::WARN))
      return false;

  return true;
}

YAML::Node ROSPackageWidget::dump()
{
  YAML::Node node(YAML::NodeType::Map);

  node[pardir_->name()] = pardir_->getValue();
  node[tbs_name_->name()] = tbs_name_->getValue();

  return node;
}

void ROSPackageWidget::load(const YAML::Node& node)
{
  pardir_->setValue(node[pardir_->name()].as<QString>());
  tbs_name_->setValue(node[tbs_name_->name()].as<QString>());
}

QString ROSPackageWidget::tbsName() const
{
  return tbs_name_->getValue();
}

QString ROSPackageWidget::tbsPath() const
{
  return tbs_path_->text();
}

void ROSPackageWidget::onPathChanged()
{
  const auto pardir = pardir_->getValue().toStdString();
  const auto tbs_name = tbs_name_->getValue().toStdString();

  const auto path = fs::path(pardir) / (tbs_name + tobas::kTBSExtension);
  tbs_path_->setText(QString::fromStdString(path));

  generate_button_->setEnabled(!pardir.empty() && !tbs_name.empty());
}

void ROSPackageWidget::onGenerateButtonClicked()
{
  Q_EMIT generateButtonClicked();
}
}  // namespace setup_assistant
}  // namespace gui
